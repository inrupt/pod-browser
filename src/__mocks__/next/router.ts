export default {
  push: jest.fn().mockResolvedValue(null),
};
