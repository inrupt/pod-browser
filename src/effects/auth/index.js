/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useEffect } from "react";
import Router from "next/router";
import { getSolidDatasetWithAcl, hasResourceAcl } from "@inrupt/solid-client";
import { useSession } from "@inrupt/solid-ui-react";
import useAuthenticatedProfile from "../../hooks/useAuthenticatedProfile";
import usePodRoot from "../../hooks/usePodRoot";

export const SESSION_STATES = {
  LOGGED_IN: "LOGGED_IN",
  LOGGED_OUT: "LOGGED_OUT",
};

export async function redirectBasedOnSessionState(
  sessionIsLoggedIn,
  sessionRequestInProgress,
  redirectIfSessionState,
  location
) {
  if (sessionRequestInProgress) {
    return;
  }

  if (
    sessionIsLoggedIn &&
    redirectIfSessionState === SESSION_STATES.LOGGED_IN
  ) {
    await Router.push(location);
  }

  if (
    !sessionIsLoggedIn &&
    redirectIfSessionState === SESSION_STATES.LOGGED_OUT
  ) {
    await Router.push(location);
  }
}

export function useRedirectIfLoggedOut(location = "/login") {
  const { session, sessionRequestInProgress } = useSession();

  useEffect(() => {
    redirectBasedOnSessionState(
      session.info.isLoggedIn,
      sessionRequestInProgress,
      SESSION_STATES.LOGGED_OUT,
      location
    );
  });
}

export function useRedirectIfLoggedIn(location = "/") {
  const { session, sessionRequestInProgress } = useSession();

  useEffect(() => {
    redirectBasedOnSessionState(
      session.info.isLoggedIn,
      sessionRequestInProgress,
      SESSION_STATES.LOGGED_IN,
      location
    );
  });
}

export function useRedirectIfNoControlAccessToOwnPod(
  resourceUrl,
  location = "/access-required"
) {
  const { session, sessionRequestInProgress } = useSession();
  const profile = useAuthenticatedProfile();
  const podRoot = usePodRoot(resourceUrl, profile);

  useEffect(() => {
    if (sessionRequestInProgress || !profile) return;

    Promise.all(
      profile.pods
        .filter((pod) => pod === podRoot)
        .map((pod) =>
          getSolidDatasetWithAcl(pod, {
            fetch: session.fetch,
          })
        )
    ).then((podResources) => {
      const hasControlAccessToPods = !podResources.find(
        (podResource) => !hasResourceAcl(podResource)
      );
      if (hasControlAccessToPods) return;
      Router.push(location);
    });
  }, [session, sessionRequestInProgress, profile, location, podRoot]);
}
