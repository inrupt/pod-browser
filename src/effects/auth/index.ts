import { useEffect } from 'react';
import Router from 'next/router';

import { getSession } from '../../../lib/solid-auth-fetcher/dist';


// TODO figure out typescript enums
export const SESSION_STATES = {
  LOGGED_IN: 'LOGGED_IN',
  LOGGED_OUT: 'LOGGED_OUT',
};

export async function redirectBasedOnSessionState(
  redirectIfSessionState: string,
  location: string,
  getSAFSession: Function,
) {
  const session = await getSAFSession();

  if (!session && redirectIfSessionState === SESSION_STATES.LOGGED_OUT) {
    Router.push(location);
  }

  if (session && redirectIfSessionState === SESSION_STATES.LOGGED_IN) {
    Router.push(location);
  }
}

export function useRedirectIfLoggedOut(
  location = '/login',
  getSAFSession = getSession,
) {
  return useEffect(() => {
    redirectBasedOnSessionState(SESSION_STATES.LOGGED_OUT, location, getSAFSession);
  }, [location, getSAFSession]);
}

export function useRedirectIfLoggedIn(
  location = '/',
  getSAFSession = getSession,
) {
  return useEffect(() => {
    redirectBasedOnSessionState(SESSION_STATES.LOGGED_IN, location, getSAFSession);
  }, [location, getSAFSession]);
}
