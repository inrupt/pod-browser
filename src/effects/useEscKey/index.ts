import { useEffect } from "react";

export const ESC_KEYCODE = 27;

export function handleEscKeydown(
  callback: (evnt: KeyboardEvent) => unknown
): (evnt: KeyboardEvent) => unknown {
  return (e: KeyboardEvent) => {
    if (!e?.keyCode) return;
    if (e.keyCode === ESC_KEYCODE) callback(e);
  };
}

export default function useEscKey(
  callback: (evnt: KeyboardEvent) => unknown
): void {
  return useEffect(() => {
    document.body.addEventListener("keydown", handleEscKeydown(callback));

    return () => {
      document.body.removeEventListener("keydown", handleEscKeydown(callback));
    };
  });
}
