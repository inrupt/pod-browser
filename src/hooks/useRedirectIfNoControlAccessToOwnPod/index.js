/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useSession } from "@inrupt/solid-ui-react";
import { useEffect, useState } from "react";
import { getResourceInfoWithAcl, hasResourceAcl } from "@inrupt/solid-client";
import Router from "next/router";
import usePodRoot from "../usePodRoot";
import useAuthenticatedProfile from "../useAuthenticatedProfile";

export default function useRedirectIfNoControlAccessToOwnPod(
  resourceUrl,
  location = "/access-required"
) {
  const { fetch, sessionRequestInProgress } = useSession();
  const { data: profile } = useAuthenticatedProfile();
  const podRoot = usePodRoot(resourceUrl, profile);
  const [redirect, setRedirect] = useState(null);

  useEffect(() => {
    if (sessionRequestInProgress || !profile) {
      setRedirect(null);
      return;
    }

    (async () => {
      const podResources = await Promise.all(
        profile.pods
          .filter((pod) => pod === podRoot)
          .map((pod) => getResourceInfoWithAcl(pod, { fetch }))
      );
      const hasControlAccessToPods = !podResources.find(
        (podResource) => !hasResourceAcl(podResource)
      );
      if (hasControlAccessToPods) {
        setRedirect(false);
        return;
      }
      await Router.push(location);
      setRedirect(true);
    })();
  }, [sessionRequestInProgress, profile, location, podRoot, fetch]);

  return redirect;
}
