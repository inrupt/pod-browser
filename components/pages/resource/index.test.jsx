/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import { useRouter } from "next/router";

import { useRedirectIfLoggedOut } from "../../../src/effects/auth";
import IndexPage from "./index";
import useRedirectIfNoControlAccessToOwnPod from "../../../src/hooks/useRedirectIfNoControlAccessToOwnPod";

jest.mock("../../../src/effects/auth");
jest.mock("../../../src/hooks/useRedirectIfNoControlAccessToOwnPod");
jest.mock("next/router");
jest.mock("../../../src/hooks/useRedirectIfNoControlAccessToOwnPod");

describe("Index page", () => {
  beforeEach(() => {
    useRouter.mockImplementation(() => ({
      query: {
        iri: encodeURIComponent("https://mypod.myhost.com"),
      },
    }));
  });

  test("Renders a logout button", () => {
    const tree = shallow(<IndexPage />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  test("Redirects if the user is logged out", () => {
    shallow(<IndexPage />);
    expect(useRedirectIfLoggedOut).toHaveBeenCalled();
  });

  test("Redirects if the user does not have access to Pod", () => {
    shallow(<IndexPage />);
    expect(useRedirectIfNoControlAccessToOwnPod).toHaveBeenCalled();
  });
});
