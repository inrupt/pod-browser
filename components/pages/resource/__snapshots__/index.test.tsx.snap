// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page Renders a logout button 1`] = `
<WithStyles(ForwardRef(Container))>
  <PodLocationProvider
    currentUri="https://mypod.myhost.com"
  >
    <DetailsMenuProvider>
      <Container
        iri="https://mypod.myhost.com"
      />
    </DetailsMenuProvider>
  </PodLocationProvider>
</WithStyles(ForwardRef(Container))>
`;
